src/vsr/replica.zig

```diff
@@ -6768,6 +7410,375 @@ pub fn ReplicaType(
+            assert(@as(std.meta.Tag(SyncStage), state_new) != state_old or
+                state_new == .request_target);
+            self.sync_stage = state_new;
+
+            // We were already syncing, but discovered a newer target:
+            if (self.sync_stage.target()) |target| {
+                assert(self.commit_stage == .idle);
+
+                if (target.checkpoint_op < self.sync_target_max.?.checkpoint_op) {
+                    log.debug("{}: sync_dispatch: change target op={}..{} id={x:0>32}..{x:0>32}", .{
+                        self.replica,
+                        target.checkpoint_op,
+                        self.sync_target_max.?.checkpoint_op,
+                        target.checkpoint_id,
+                        self.sync_target_max.?.checkpoint_id,
+                    });
+                    self.sync_stage = .{ .request_trailers = .{ .target = self.sync_target_max.? } };
+                }
+            }
+
+            // We have been waiting for a target:
+            if (self.sync_stage == .request_target) {
+                assert(self.commit_stage == .idle);
+
+                if (self.sync_target_max) |target| {
+                    if (target.checkpoint_op > self.op_checkpoint() or
+                        (target.checkpoint_op == self.op_checkpoint() and
+                        target.checkpoint_id != self.superblock.working.checkpoint_id()))
+                    {
+                        self.sync_stage = .{ .request_trailers = .{ .target = target } };
+                    }
+                }
+            }
+
+            if (self.sync_stage.target()) |target| {
+                log.debug("{}: sync_dispatch: {s}..{s}: " ++
+                    "(target_op={} target_id={x:0>32})", .{
+                    self.replica,
+                    @tagName(state_old),
+                    @tagName(self.sync_stage),
+                    target.checkpoint_op,
+                    target.checkpoint_id,
+                });
+            } else {
+                log.debug("{}: sync_dispatch: {s}..{s}", .{
+                    self.replica,
+                    @tagName(state_old),
+                    @tagName(self.sync_stage),
+                });
+            }
+
+            if (self.test_event_callback) |hook| hook(self, .sync);
+
+            switch (self.sync_stage) {
+                .none => unreachable,
+                .cancel_commit => {}, // Waiting for an uninterruptible commit step.
+                .cancel_grid => self.grid.cancel(sync_cancel_grid_callback),
+                .request_target => {}, // Waiting for a usable sync target.
+                .request_trailers => self.sync_message_timeout.start(),
+                .write_sync_start => self.sync_superblock_start(),
+                .request_manifest_logs => self.sync_request_manifest_logs(),
+                .write_sync_done => self.sync_superblock_done(),
+                .done => {
+                    self.sync_stage = .none;
```

Comment:
can we assert relations between `.op`, `.commit_min` and `status == .recovering_head` here? Or maybe nothere, but upon the entry of dispatch function --- my reading is taht `.recovering_head` is actually normal case here, but the switch to that state is burried in `sync_superblock_start_callback.`